lib/chef/provider/aws_iam_instance_profile.rb

```diff
@@ -0,0 +1,49 @@
+require 'chef/provisioning/aws_driver/aws_provider'
+
+class Chef::Provider::AwsInstanceProfile < Chef::Provisioning::AWSDriver::AWSProvider
+  provides :aws_iam_instance_profile
+
+  def action_create
+    iam_instance_profile = super
+
+    if new_resource.role && !iam_instance_profile.roles.map(&:name).include?(new_resource.role)
```

Comment:
I'm fairly certain that this logic will need to be embedded into a `update_attached_role(iam_instance_profile)` method and called from both `action_create` and `update_aws_object`.